### ISSUE TITLE
Torch Cpu Memory Leak with FastApi uvicorn.

### ISSUE THREAD

**Issue by ketangangal** [pytorch/pytorch#114455]
### 🐛 Describe the bug

**Context :** 
I am using fastapi to expose embeddings model. Code contains torch.no_grad() , gc, and del model. 
But still during embeddings generation 250-300 mb memory getting stuck in ram. 

![image](https://github.com/pytorch/pytorch/assets/40850370/0c2e3b6a-00ac-45e4-9200-169f775c782d)

Similar issue: https://github.com/pytorch/pytorch/issues/68114

Is there any way to release cpu memory after every api call ? 

### Versions

**Context :** 
I am using fastapi to expose embeddings model. Code contains torch.no_grad() , gc, and del model. 
But still during embeddings generation 250-300 mb memory getting stuck in ram. 

![image](https://github.com/pytorch/pytorch/assets/40850370/0c2e3b6a-00ac-45e4-9200-169f775c782d)

Similar issue: https://github.com/pytorch/pytorch/issues/68114

Is there any way to release cpu memory after every api call ?

**Comment by malfet** [comment]
Need larger code example to understand what you are doing, but frankly I don't see the problem: `generator.generate` might allocate memory for some internal structure associated with the object, and while there is a reference to an object is memory is in use, but it should go down once you release the object (for example by issuing `del a`)

**Comment by ketangangal** [comment]
Hello @malfet , 

At line number 89 it keeps on adding 100 mb approx for each batch !

![image](https://github.com/pytorch/pytorch/assets/40850370/365fddac-ea59-4266-8862-1007e766a2e2)

**Comment by ketangangal** [comment]
If you observe line number 111, after putting outputs to None it releases 48 mb.  
ps: i was doing some random exp with this to figure out root cause.

**Comment by ketangangal** [comment]
Hi @malfet did you get time to look at this issue, do i need to provide anything else ?

**Comment by MvMukesh** [comment]
@ketangangal have you tried " torch.cuda.empty_cache() "

**Comment by Meshwa428** [comment]
any success in finding out the issue?
mine seems to be happening while using diffusers library.

after each image generation the memory usage keeps increasing.
like after warmup it sits at 67% then after each image generation it increases about 3%
first 67%
then goes up to 70% and sits there
then 73%
sits there and then similary till 95% and then if i generate another image it just does oom

**Comment by Meshwa428** [comment]
Does that work on windows?
I tried porting some jemalloc but it didn't work 😔

**Comment by Elliott-Song** [comment]
@cbensimon Thanks! Your ChatGPT excerpt had the answer for me:

```
import ctypes
from ctypes import wintypes

# Define necessary Windows types and constants
SetProcessWorkingSetSize = ctypes.windll.kernel32.SetProcessWorkingSetSize
GetCurrentProcess = ctypes.windll.kernel32.GetCurrentProcess

# SetProcessWorkingSetSize function prototype
SetProcessWorkingSetSize.argtypes = [wintypes.HANDLE, ctypes.c_size_t, ctypes.c_size_t]
SetProcessWorkingSetSize.restype = wintypes.BOOL

# Call the function to reduce the working set size
def release_unused_memory():
    current_process = GetCurrentProcess()
    success = SetProcessWorkingSetSize(current_process, -1, -1)
    if not success:
        raise ctypes.WinError()

release_unused_memory()
```
